Add Telegram scheduling implementation

azzilian opened the pull request:
- Update CarSharingServiceApplication.java - scheduleenabled
- Update RentalRepository.java - method with query added to gather all with date tomorrow or less
- Add NotificationScheduler & NotificationSchedulerImpl - schedule for morning 08:30 and evening 16:30 notifications + gathering overdued rentals
- Add NotificationServiceImpl.java - added sendNotifications
- Refactor NotificationServiceImpl.java - sendMessageToAllChats now separate method!

nklimovych committed:
Initial commit

azzilian committed:
AppConfig.java - added Bean for BotController for registration
BotService.java - register user + send message about user
Liquibase - raw creation table + users insert
UserRepository - existByChatId + findByChatId
UserService - search user in DB + search chatId
Raw User entity - need to update
pom.xml - telegram dependency added

azzilian committed:
AppConfig.java - fixes to be able to read enviroment variables
BotService.java - handlers added

azzilian committed:
Refactor - AppConfig.java -> BotConfig + code change
Add - NotificationService.java
Refactor - NotificationServiceImpl.java - added new functional
Refactor - db.changelog-master.yaml - stub
Add - Rental,RentalService - stub
Removed - liquibase users creation stubs

SkillfulDev committed:
Add - Authentication with JWT

azzilian committed:
Refactor - RentalRepository.java - stub and querry for returnDate
Refactor - UserRepository - stub
Refactor - pom.xml - EV now could be gained from .env
Refactor - NotificationServiceImpl - now takes EV from .env file + new logic to send Status info about returnDate
Remove - application.properties - EV removed

azzilian committed:
Refactor - NotificationServiceImpl - reading EV from application properties
Remove - pom.xml - revomed dependency for reading from .env

SkillfulDev committed:
Refactor - change RoleName to RoleType

azzilian committed:
Remove - Rental completely
Add - Car with stubs for controller
Refactor - NotificationService added method with carCreation with stub

azzilian committed:
Add - UserChatId - entity to connect users and chats
BugFix - CarRepository - correct type returns
Remove - CarController - authoriation removed - DO NOT MERGE
Refactor - NotificationServiceImpl.java -now send message to all users with chats with bot

azzilian committed:
Bugfix - empty line removed + non-used peendency removed

azzilian committed:
Removed command /returStatus - will be added in other task

azzilian committed:
Update - NotificationServiceImpl.java - response style update

azzilian committed:
Add - test bot for est enviroment

azzilian committed:
Added Stub classes for Rental creation

azzilian committed:
Added find by Id for user
Refactor NotificationServiceImpl - proper email responce

azzilian committed:
Removed stubs
Add RentalRepository
Refactor RentalController
Add CarsNotAvailableException
Refactor File Structure
Refactor Mapper - responsedto

azzilian committed:
Update RentalRepository - findOverdueRentals wit Query
Add NotificationScheduler and NotificationServiceImpl - send reminder twice in the morning and evening (start and end of the shift)
Update CarSharingServiceApplication - enable scheduling
Refactor NotificationService - added method for notifications separated method to sendToAll

azzilian committed:
Refactor - NotificationSchedulerImp - Strings moved  to constants failed to set them in properties

azzilian committed:
BugFix - day changed to tomorrow according to task

azzilian committed:
BugFix - day changed to tomorrow according to task

azzilian committed:
Merge branch main into telegram-implementation

azzilian committed:
Refactor RentalRequestDto - Removed @NotBlack

azzilian committed:
Merge remote-tracking branch 'origin/telegram-implementation' into telegram-scheduling-implementation

# Conflicts:
#	src/main/java/online/carsharing/repository/rental/RentalRepository.java

azzilian committed:
Merge branch telegram-implementation into telegram-scheduling-implementation
Refactor - RentalRepository - Removed @EntityGraph Left Join Fetch used instead

azzilian committed:
Merge remote-tracking branch 'origin/main' into telegram-scheduling-implementation

# Conflicts:
#	src/main/java/online/carsharing/dto/request/rental/RentalRequestDto.java
#	src/main/java/online/carsharing/entity/Car.java
#	src/main/java/online/carsharing/repository/rental/RentalRepository.java
#	src/main/java/online/carsharing/service/NotificationService.java
#	src/main/java/online/carsharing/service/impl/CarServiceImpl.java
#	src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java
#	src/main/java/online/carsharing/service/impl/RentalServiceImpl.java
#	src/main/resources/db/changelog/changes/vyshynskyi/create-users-chat_id-table.yaml
#	src/main/resources/db/changelog/db.changelog-master.yaml

azzilian committed:
Merge main into telegram-scheduling-implementation

Refactor:
 application.properties - BOTNAME changed to BOT_NAME, BOTTOKEN changed to BOT_TOKEN
 NotificationServiceImpl - order to createCarNotification - brand,model,type,fee
Removed:
NotificationSchedulerImp - RentalResponseDto no longer used

azzilian committed:
Refactor Merge conflicts resolve

azzilian committed:
Refactor UserChatIdRepository now message will be be send only to one chat despite of number of users in this chat
NotificationSchedulerImp - logger removed

nklimovych commented on src/main/java/online/carsharing/repository/rental/RentalRepository.java:
```diff
@@ -12,4 +14,8 @@ public interface RentalRepository extends JpaRepository<Rental, Long> {
+
+    @Query("SELECT r FROM Rental r LEFT JOIN FETCH r.car LEFT JOIN FETCH"
+            + " r.user WHERE r.returnDate <= :tomorrow AND r.actualReturnDate IS NULL")
+    List<Rental> findOverdueRentals(@Param("tomorrow") LocalDate tomorrow);
```
It is not necessary to write `@Param` before the `tomorrow` parameter. I tested this method it works fine.
```suggestion

    @Query("SELECT r FROM Rental r LEFT JOIN FETCH r.car LEFT JOIN FETCH"
            + " r.user WHERE r.returnDate <= :tomorrow AND r.actualReturnDate IS NULL")
    List<Rental> findOverdueRentals(LocalDate tomorrow);
```

nklimovych commented on src/main/java/online/carsharing/service/impl/NotificationSchedulerImp.java:
```diff
@@ -0,0 +1,76 @@
+
+@Service
+@RequiredArgsConstructor
+public class NotificationSchedulerImp implements NotificationScheduler {
```
```suggestion
public class NotificationSchedulerImpl implements NotificationScheduler {

```

nklimovych commented on src/main/java/online/carsharing/service/impl/NotificationSchedulerImp.java:
```diff
@@ -0,0 +1,76 @@
+                String message = String.format(EXPIRED_RENTALS,
+                        rental.getRentalDate(), rental.getReturnDate(),
+                        carModel, userEmail);
+                notificationService.sendNotification(message);
```
i'm not sure, but i think it looks better (it's up to you)
```suggestion
                String message = String.format(
                        EXPIRED_RENTALS,
                        rental.getRentalDate(), 
                        rental.getReturnDate(),
                        carModel, 
                        userEmail);
                notificationService.sendNotification(message);

```

nklimovych commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -95,31 +95,16 @@ public void sendMessage(Update update, String text) {
-                            .parseMode("Markdown")
-                            .build();
-                    execute(message);
-                } catch (TelegramApiException e) {
-                    e.printStackTrace();
-                }
-            }
-        }
+        sendMessageToAllChats(messageText);
```
same here

nklimovych replied:
I selected lines 99-102

azzilian replied:
done, btw 103 - 117 fixed in PR#17

nklimovych commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -129,6 +114,16 @@ public void createRentalNotification(Rental rental) {
+    }
+
+    private void sendMessageToAllChats(String messageText) {
+        List<Long> chatIds = userChatIdRepository.findAllChatIds();
```
think about it, maybe it make sense:
```suggestion
        List<Long> chatIds = chatRepository.findAll();

```

azzilian replied:
Method findAllChatId should stay because of changs to table
chaRepository - ok I can do that

nklimovych reviewed (commented):
Nothing critical, left some suggestions and comments for you

azzilian committed:
Refactor:
UserChatId - primary key changed to new field id
NotificationSchedulerImpl - renamed + code refactor
RentalRepository - @Param removed
UserChatIdRepository - method findByUserId renamed to findUserByTelegramId

SkillfulDev commented on src/main/java/online/carsharing/service/impl/NotificationSchedulerImpl.java:
```diff
@@ -0,0 +1,82 @@
+        rentalReminder();
+    }
+
+    private void rentalReminder() {
+        LocalDate tomorrow = LocalDate.now().plusDays(DAY);
+        List<Rental> overdueRentals = rentalRepository.findOverdueRentals(tomorrow);
+
+        if (overdueRentals.isEmpty()) {
+            notificationService.sendNotification(NO_RENTALS);
+        } else {
+            notificationService.sendNotification(RENTAL_WARNING_TG_HEADER);
+            for (Rental rental : overdueRentals) {
+                String carModel = carRepository.findById(rental
+                                .getCar()
+                                .getId())
+                        .map(car -> car.getModel())
+                        .orElse("Unknown Car");
+                String userEmail = userRepository.findById(rental
+                                .getUser()
+                                .getId())
+                        .map(user -> user.getEmail())
+                        .orElse("Unknown User");
+                String message = String.format(
+                        EXPIRED_RENTALS,
+                        rental.getRentalDate(),
+                        rental.getReturnDate(),
+                        carModel, userEmail);
+                notificationService.sendNotification(message);
+            }
+            notificationService.sendNotification(RENTAL_WARNING_TG_TOTAL_OVERDUE
+                    + overdueRentals.size());
+        }
```
I would consider splitting the method

nklimovych replied:
I agree

azzilian replied:
Ok I will separate this in PR 17
